lib/Parse/ParseDecl.cpp

```diff
@@ -1511,9 +1511,10 @@ bool Parser::parseNewDeclAttribute(DeclAttributes &Attributes, SourceLoc AtLoc,
     InlineKind kind;
     if (Tok.getText() == "never")
       kind = InlineKind::Never;
-    else if (Tok.getText() == "__always")
+    else if (Tok.getText() == "__always") {
       kind = InlineKind::Always;
-    else {
+      if (Context.LangOpts.IgnoreAlwaysInline) DiscardAttribute = true;
```

Comment:
I'm a bit surprised that you're doing this in the parser. I suspect you'd be better off checking this in SILGen, where the attribute gets added to SIL functions, so that it only affects code generation.

Reply:
That was my original approach too, but then I realized that the attribute would then still "leak" into declarations. I had a change in SILDeclRef::isAlwaysInline to return false, but the testcase I've added would emit this:
```
sil_stage canonical
import Builtin
import Swift
import SwiftShims
@inline(__always) func foo() // <=== always_inline still here
sil hidden @$s4main3fooyyF : $@convention(thin) () -> () { // <=== always_inline removed
...
}
```

I am not sure what the consequences of that is. Would that be a better solution?